miner/verify_bundles.go

```diff
@@ -0,0 +1,285 @@
+}
+
+type bundleTxData struct {
+	hash      common.Hash
+	canRevert bool
+}
+
+type includedTxData struct {
+	hash     common.Hash
+	index    int
+	reverted bool
+}
+
+type privateTxData struct {
+	bundleHash common.Hash
+	index      int
+}
+
+// checkBundlesAtomicity checks that all txs from the included bundles are included in the block correctly
+// 1. We check that all non-reverted txs from the bundle are included in the block and are not reverted
+// 2. Reverted txs are allowed to be not included in the block
+// 3. All txs from the bundle must be in the right order, gaps between txs are allowed
+func checkBundlesAtomicity(
+	includedBundles map[common.Hash][]bundleTxData,
+	includedTxDataByHash map[common.Hash]includedTxData,
+	privateTxsFromFailedBundles map[common.Hash]privateTxData,
+) error {
+	txsFromSuccessfulBundles := make(map[common.Hash]struct{})
+
+	for bundleHash, b := range includedBundles {
+		var (
+			firstTxBlockIdx  int
+			firstTxBundleIdx int
+		)
+		// 1. locate the first included tx of the bundle
+		for bundleIdx, tx := range b {
+			txsFromSuccessfulBundles[tx.hash] = struct{}{}
+
+			txInclusion, ok := includedTxDataByHash[tx.hash]
+			if !ok {
+				// tx not found, maybe it was reverting
+				if tx.canRevert {
+					continue
+				} else {
+					return NewErrBundleTxNotFound(bundleHash, tx.hash, bundleIdx)
+				}
+			}
+
+			if txInclusion.reverted && !tx.canRevert {
+				return NewErrBundleTxReverted(bundleHash, tx.hash, bundleIdx)
+			}
+
+			firstTxBlockIdx = txInclusion.index
+			firstTxBundleIdx = bundleIdx
+			break
+		}
+
+		currentBlockTx := firstTxBlockIdx + 1
+		// locate other txs in the bundle
+		for idx, tx := range b[firstTxBundleIdx+1:] {
+			txsFromSuccessfulBundles[tx.hash] = struct{}{}
+
+			bundleIdx := firstTxBundleIdx + 1 + idx
+			// see if tx is on its place
+			txInclusion, ok := includedTxDataByHash[tx.hash]
+			if !ok {
+				// tx was not found, maybe its reverting
+				if tx.canRevert {
+					continue
+				} else {
+					return NewErrBundleTxNotFound(bundleHash, tx.hash, bundleIdx)
+				}
+			}
+
+			// we allow gaps between txs in the bundle,
+			// but txs must be in the right order
+			if txInclusion.index < currentBlockTx {
+				return NewErrBundleTxWrongPlace(bundleHash, tx.hash, bundleIdx, txInclusion.index, currentBlockTx)
+			}
+
+			if txInclusion.reverted && !tx.canRevert {
+				return NewErrBundleTxReverted(bundleHash, tx.hash, bundleIdx)
+			}
+
+			currentBlockTx = txInclusion.index + 1
+		}
+	}
+
+	for hash, priv := range privateTxsFromFailedBundles {
+		if _, ok := txsFromSuccessfulBundles[hash]; ok {
+			continue
+		}
+		if _, ok := includedTxDataByHash[hash]; ok {
+			return NewErrPrivateTxFromFailedBundle(priv.bundleHash, hash, priv.index)
+		}
+	}
+
+	return nil
+}
+
+func extractBundleTxDataFromBundles(bundles []types.SimulatedBundle, result map[common.Hash][]bundleTxData) {
```

Comment:
Minor suggestion: I'd expect result as an outparam to be a pointer to the map because I never know if the map is a reference under the copied value

Reply:
I agree but pointer to map does not support `m[key]` operation so you would still need to do `(*m)[key]` below.

Reply:
I think that's fine